docs/aztec/protocol/trees/indexed-merkle-tree.mdx

```diff
@@ -3,66 +3,71 @@ title: Indexed Merkle Tree
 
-4. Add a new value $v=20$
+## Indexed Merkle Tree Constructions
 
-<Image img={require("/img/index-merkle-tree-4.png")} />
+As it turns out, we can do better. [This paper](https://eprint.iacr.org/2021/1263.pdf) (page 6) introduces the idea of an indexed merkle tree. A Merkle Tree that allows us to perform efficient non-membership proofs. It achieves this by extending each node to include a specialized data structure. Each node not only stores some value $v \in \mathbb{F}_p$ but also some pointers to the leaf with the next higher value. The data structure is as follows:
 
-5. Add a new value $v=50$
+$$
+\textsf{leaf} = \{v, i_{\textsf{next}}, v_{\textsf{next}}\}.
+$$
 
-<Image img={require("/img/index-merkle-tree-5.png")} />
+Based on the tree's insertion rules, we can assume that there are no leaves in the tree that exist between the range $(v, v_{\textsf{next}})$.
 
-## Analysis - Constraint counts
+More simply put, the merkle tree pretty much becomes a linked list of increasing size, where once inserted the pointers at a leaf can change, but the nullifier value cannot.
 
-### Insertion
+Despite being a minor modification, the performance implications are massive. We no longer position leaves in place $(index == value)$ therefore we no longer need a deep tree, rather we can use an arbitrarily small tree (32 levels should suffice). Some quick back of the napkin maths can show that insertions can be improved by a factor of 8 $(256 / 32)$.
 
-- Hash the old leaf: `oldLeaf = h(0, 0, 0)` (for example).
-- Prove the old leaf existed in the tree: `n` hashes.
-- Check the new leaf belongs in the slot it's being inserted into: `2` range checks.
-  - If (`oldLeaf.nextInd == 0`):
-    - Special case, we're inserting at the very end, so the new leaf must be the highest-valued so far.
-    - `assert(oldLeaf.val < newLeaf.val)`
-  - Else:
-    - `assert(oldLeaf.val < newLeaf.val)`
-    - `assert(oldLeaf.nextVal > newLeaf.val)`
-- Copy over the old leaf's `next` values to the new leaf:
-  - `assert(oldLeaf.nextInd == newLeaf.nextInd)`
-  - `assert(oldLeaf.nextVal == newLeaf.nextVal)`
-- Update the old leaf to point to the new leaf: `updatedLeaf = h(0, 30, 1)`.
-  - `assert(updatedLeaf.nextInd == newLeaf.ind)`
-  - `assert(updatedLeaf.nextVal == newLeaf.val)`
-- Replace the `updatedLeaf` in the tree: `n` hashes.
-- Hash the new leaf: `newLeaf = h(30, 0, 0)`.
-- Add the `newLeaf` to the tree: `n` hashes.
+_For the remainder of this article I will refer to the node that provides the non membership check as a "low nullifier"._
+
+The insertion protocol is described below:
+
+1. Look for a nullifier's corresponding low*nullifier, where $low_nullifier*{\textsf{next_value}} > v$.
```

Comment:
ah good spot i didnt notice this

Reply:
On this, should we add a note that it might be the largest so if there is no value that have `low\_nullifier_{\mathsf{next\_value}} > v` going for the 0 is the way to go? We are doing the check to accept 0 later, but not mentioning that it is ok.